### PR TITLE
Challenge Handler for Embedded Webview

### PR DESCRIPTION
**Pull request by jasoncoolmax**
PKeyAuth change also included.

### CODE REVIEW

File: IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m [hunk 1]
```diff
@@ -499,6 +525,36 @@ - (UIActivityIndicatorView *)prepareLoadingIndicator:(UIView *)rootView
     return loadingIndicator;
 }
 
+- (void)handlePKeyAuthChallenge:(NSString *)challengeUrl
+{
```
Comment (jasoncoolmax) [inline IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m]: Discussion:
I am thinking about putting such a function to `MSIDPKeyAuthHandler`, although such a challenge is not caught the same way as other handlers? Thoughts? :)

File: IdentityCore/src/util/MSIDHelpers.m [hunk 2]
```diff
@@ -22,6 +22,8 @@
 // THE SOFTWARE.
 
 #import "MSIDHelpers.h"
+#import "MSIDContants.h"
```
Comment (unpluggedk) [inline IdentityCore/src/util/MSIDHelpers.m]: just realized we had a typo here.
Constants perhaps

File: IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m [hunk 3]
```diff
@@ -382,7 +405,10 @@ - (void)webView:(WKWebView *)webView didReceiveAuthenticationChallenge:(NSURLAut
                      @"session:task:didReceiveChallenge:completionHandler",
```
Comment (unpluggedk) [inline IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m]: session:task:didReceiveChallenge:completionHandler

File: IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m (hunk 1 above)
Comment (unpluggedk) [reply IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m]: Yes, I think you should.

File: IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m [hunk 4]
```diff
@@ -499,6 +525,36 @@ - (UIActivityIndicatorView *)prepareLoadingIndicator:(UIView *)rootView
...
+{
+    MSID_LOG_INFO(_context, @"Handling PKeyAuth Challenge.");
+    
+    NSArray * parts = [challengeUrl componentsSeparatedByString:@"?"];
+    NSString *qp = [parts objectAtIndex:1];
+    NSDictionary* queryParamsMap = [NSDictionary msidURLFormDecode:qp];
+    NSString* submitUrl = [MSIDHelpers msidAddClientVersionToURLString:[queryParamsMap valueForKey:@"SubmitUrl"]];
+    
+    NSArray * authorityParts = [submitUrl componentsSeparatedByString:@"?"];
+    NSString *authority = [authorityParts objectAtIndex:0];
```
Comment (unpluggedk) [inline IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m]: nit: * through out placed inconsistantly

File: IdentityCore/src/workplacejoin/MSIDPkeyAuthHelper.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import <Foundation/Foundation.h>
+#import "MSIDRegistrationInformation.h"
+
+@interface MSIDPkeyAuthHelper : NSObject
+
++ (nullable NSString *)createDeviceAuthResponse:(nonnull NSString *)authorizationServer
+                                  challengeData:(nullable NSDictionary *)challengeData
```
Comment (unpluggedk) [inline IdentityCore/src/workplacejoin/MSIDPkeyAuthHelper.h]: would we ever call this without a challengeData?

File: IdentityCore/src/webview/embeddedWebview/challangeHandlers/ui/mac/MSIDCredentialCollectionController.m [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDCredentialCollectionController.h"
+
+// UI position values for the input dialog
+const CGFloat USERNAME_LABEL_X = 7;
```
Comment (unpluggedk) [inline IdentityCore/src/webview/embeddedWebview/challangeHandlers/ui/mac/MSIDCredentialCollectionController.m]: Why not create a xib ?

File: IdentityCore/src/webview/embeddedWebview/challangeHandlers/ui/mac/MSIDNTLMUIPrompt.m [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        MSIDCredentialCollectionController *view = [MSIDCredentialCollectionController new];
+        [view.usernameLabel setStringValue:NSLocalizedString(@"Username", nil)];
+        [view.passwordLabel setStringValue:NSLocalizedString(@"Password", nil)];
+        [alert setAccessoryView:view.customView];
+        
+        [[alert window] setInitialFirstResponder:view.usernameField];
+        
+        [alert beginSheetModalForWindow:[NSApp keyWindow] completionHandler:^(NSModalResponse returnCode)
+         {
+             if (returnCode == 1000)
```
Comment (unpluggedk) [inline IdentityCore/src/webview/embeddedWebview/challangeHandlers/ui/mac/MSIDNTLMUIPrompt.m]: What is 1000?

File: IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m (hunk 4 above)
Comment (jasoncoolmax) [reply IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m]: done

File: IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m (hunk 1 above)
Comment (jasoncoolmax) [reply IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m]: done

File: IdentityCore/src/workplacejoin/MSIDPkeyAuthHelper.h (hunk 5 above)
Comment (jasoncoolmax) [reply IdentityCore/src/workplacejoin/MSIDPkeyAuthHelper.h]: Only if server sends us a PKeyAuth challenge without challenge data (I have never seen that before :)  ).

I have just added the logging for it if that happens.

File: IdentityCore/src/webview/embeddedWebview/challangeHandlers/ui/mac/MSIDCredentialCollectionController.m (hunk 6 above)
Comment (jasoncoolmax) [reply IdentityCore/src/webview/embeddedWebview/challangeHandlers/ui/mac/MSIDCredentialCollectionController.m]: If I remember correctly, previously we were using xib/nib, then changed to storyboard. As those things are always changing and evolving, finally decided to remove all xib/nib/storyboard in project and build UI programmatically :)

File: IdentityCore/src/webview/embeddedWebview/challangeHandlers/ui/mac/MSIDNTLMUIPrompt.m (hunk 7 above)
Comment (jasoncoolmax) [reply IdentityCore/src/webview/embeddedWebview/challangeHandlers/ui/mac/MSIDNTLMUIPrompt.m]: changed to constant and added comment to clarify it :)

File: IdentityCore/src/util/MSIDHelpers.m (hunk 2 above)
Comment (jasoncoolmax) [reply IdentityCore/src/util/MSIDHelpers.m]: fixed

**Pull request merged**